src/tracks/ui/BrushHandle.cpp

```diff
@@ -0,0 +1,567 @@
+      return Cancelled;
+
+   wxMouseEvent &event = evt.event;
+   const auto sTrack = TrackList::Get( *pProject ).Lock( FindTrack() );
+   const auto pTrack = sTrack.get();
+   const WaveTrack *const wt =
+         static_cast<const WaveTrack*>(pTrack);
+   auto &trackPanel = TrackPanel::Get( *pProject );
+   auto &viewInfo = ViewInfo::Get( *pProject );
+
+   mMostRecentX = event.m_x;
+   mMostRecentY = event.m_y;
+
+   return RefreshAll;
+}
+
+UIHandle::Result BrushHandle::Drag
+      (const TrackPanelMouseEvent &evt, AudacityProject *pProject)
+{
+   using namespace RefreshCode;
+
+   const auto pView = mpView.lock();
+   if ( !pView )
+      return Cancelled;
+
+   wxMouseEvent &event = evt.event;
+   const auto sTrack = TrackList::Get( *pProject ).Lock( FindTrack() );
+   const auto pTrack = sTrack.get();
+   WaveTrack *const wt =
+         static_cast<WaveTrack*>(pTrack);
+   auto &trackPanel = TrackPanel::Get( *pProject );
+   auto &viewInfo = ViewInfo::Get( *pProject );
+
+   int x = mAutoScrolling ? mMostRecentX : event.m_x;
+   int y = mAutoScrolling ? mMostRecentY : event.m_y;
+   mMostRecentX = x;
+   mMostRecentY = y;
+
+   // Clip the coordinates
+   // TODO: Find ways to access the ClipParameters (for the mid)
+   int x1 = std::max(mRect.x + 10, std::min(x, mRect.x + mRect.width - 20));
+   int y1 = std::max(mRect.y + 10, std::min(y, mRect.y + mRect.height - 10));
+
+   mbCtrlDown = event.ControlDown();
+
+   if(!mpSpectralData->coordHistory.empty()){
+      int x0 = mpSpectralData->coordHistory.back().first;
+      int y0 = mpSpectralData->coordHistory.back().second;
+      int wd = mBrushRadius * 2;
+
+      int dx =  abs(x1-x0), sx = x0<x1 ? 1 : -1;
+      int dy = -abs(y1-y0), sy = y0<y1 ? 1 : -1;
+      int err = dx+dy, err2;
+
+      double posTime;
+      wxInt64 posFreq;
+
+      auto posToLongLong = [&](int x0){
+         posTime = viewInfo.PositionToTime(x0, mRect.x);
+         sampleCount sc = wt->TimeToLongSamples(posTime);
+         return sc.as_long_long();
+      };
+
+      // Line drawing (draw points until the start coordinate reaches the end)
+      while(true){
+         if (x0 == x1 && y0 == y1)
+            break;
+         err2 = err * 2;
+         if (err2 * 2 >= dy) { err += dy; x0 += sx; }
+         if (err2 * 2 <= dx) { err += dx; y0 += sy; }
+
+         // For each (x0, y0), draw circle based on Quadrant
+         int xm = x0, ym = y0;
+         int r = mBrushRadius;
+         int cx = -r, cy = 0, cerr = 2 - 2 * r;
+
+         if(mFreqSnapping){
+            // Correct the y coord (snap to highest energy freq. bin)
+            long long startSC = posToLongLong(xm);
+            wxInt64 targetFreq = PositionToFrequency(wt, 0, ym+cy, mRect.y, mRect.height);
+            if(auto *sView = dynamic_cast<SpectrumView*>(pView.get())){
+               wxInt64 resFreq = SpectralDataManager::FindFrequencySnappingBin(wt,
+                                                                               startSC,
```

Comment:
Each time you call this function, a SpectrumTransformer is created and destoyed.

If you round sample times to hop sizes, you might be computing the same FFTs repeatedly.

Maybe the SpectrumTransformer object could have a longer lifetime, and a cache of FFT windows?

This may be a problem for me to think about and implement to improve my class.